blockstore/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+#![cfg_attr(docs_rs, feature(async_fn_in_trait))]
+
+use cid::CidGeneric;
+use thiserror::Error;
+
+use crate::multihash::{Block, CidError};
+
+pub use crate::in_memory_blockstore::InMemoryBlockstore;
+
+mod in_memory_blockstore;
+/// Utilities related to computing CID for the inserted data
+pub mod multihash;
+
+/// Error returned when performing operations on [`Blockstore`]
+#[derive(Debug, PartialEq, Error)]
+pub enum BlockstoreError {
+    /// Provided CID already exists in blockstore when trying to insert it
+    #[error("CID already exists in the store")]
+    CidExists,
+
+    /// Provided CID is longer than max length supported by the blockstore
+    #[error("CID length longer that max allowed by the store")]
+    CidTooLong,
+
+    /// Error occured when trying to compute CID with [`HasCid`] trait
+    ///
+    /// [`HasCid`]: crate::multihash::HasCid
+    #[error("Error generating CID: {0}")]
+    CidError(#[from] CidError),
+}
+
+type Result<T> = std::result::Result<T, BlockstoreError>;
+
+/// A IPLD blockstore capable of holding arbitrary data indexed by CID. Implementations can impose
+/// limit on supported CID length, and any operations on longer CIDs will fail with [`CidTooLong`]
+///
+/// [`CidTooLong`]: BlockstoreError::CidTooLong
+
+#[cfg_attr(not(docs_rs), async_trait::async_trait)]
```

Comment:
I've taken this from https://github.com/dtolnay/async-trait/issues/213
otherwise generated docs show de-sugared async trait with ton of additional lifetimes.
Unfortunately, to get the same effect locally, one has to generate docs with `RUSTDOCFLAGS="--cfg docs_rs" cargo +nightly doc`, not sure if there's neater way to get the same result.

Reply:
Per daily, I've tried compiling with beta toolchain and without async-trait crate and yeah, there's a warning that using async fn in public traits is discouraged, because it doesn't generate auto traits. Seems like we could stick with the crate or follow the suggestion of desugaring the `fn` ourselves 
```
42 -     async fn get<const S: usize>(&self, cid: &CidGeneric<S>) -> Result<Option<Vec<u8>>>;
42 +     fn get<const S: usize>(&self, cid: &CidGeneric<S>) -> impl std::future::Future<Output = Result<Option<Vec<u8>>>> + Send;
```
wdyt?

Reply:
I am okay with desugaring until this is corrected in std, it is the approach suggested by async wg

Reply:
> warning that using async fn in public traits is discouraged

wait, so it is discouraged in public traits? then I'm leaning towards `async_trait` :sweat_smile:

Reply:
yeah, we might want to stick with `async_trait`. This is the full warning:
```
warning: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
  --> blockstore/src/lib.rs:42:5
   |
42 |     async fn get<const S: usize>(&self, cid: &CidGeneric<S>) -> Result<Option<Vec<u8>>>;
   |     ^^^^^
   |
   = note: you can suppress this lint if you plan to use the trait only in your own code, or do not care about auto traits like `Send` on the `Future`
   = note: `#[warn(async_fn_in_trait)]` on by default
help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`, but these cannot be relaxed without a breaking API change
   |
42 -     async fn get<const S: usize>(&self, cid: &CidGeneric<S>) -> Result<Option<Vec<u8>>>;
42 +     fn get<const S: usize>(&self, cid: &CidGeneric<S>) -> impl std::future::Future<Output = Result<Option<Vec<u8>>>> + Send;
   |
```
Plus discussion from the async fn stabilisation https://github.com/rust-lang/rust/pull/115822#issuecomment-1731149475

Reply:
@oblique

Reply:
I think I like the workaround with `docsrs` flag